Consistency between goto programs and symbol table

xbauch opened the pull request:
Introduce a chain of methods that (at the lowest level) iterate over all symbols
in a goto program and for each one check that it is present in the symbol table
as well. On the lower levels, the methods collect inconsistency messages and
print them to error output. On the top level, e.g. cbmc-parser, they return a
Boolean value to be caught by an invariant check.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis commented on src/goto-programs/goto_model.h:
```diff
@@ -18,6 +18,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "abstract_goto_model.h"
 #include "goto_functions.h"
 
+#include <iostream>
```
`#include <iostream>` in a header is a very bad idea. Please remove this and move the implementation of `check_id_consistency_symbol_table` to a `.cpp` file.

xbauch replied:
Ok, removed `iostream` and reworked to use `messaget` instance.

NlightNFotis commented on src/goto-programs/goto_model.h:
```diff
@@ -95,6 +97,29 @@ class goto_modelt : public abstract_goto_modelt
+    }
+    if(!message_collector.empty())
+    {
+      std::cerr << "Error messages collected while checking the goto "
```
Also, can you see if you have a reference to a message handler like `error()` or `info()`, instead of writing to `std::cerr`? If it's possible please do that instead.

tautschnig commented:
I'm not sure I'm convinced this is a design that is easy to extend - it seems each additional check will required about the same amount of changes? See #918 for an attempt of mine that should make it fairly easy to add new checks?

NlightNFotis reviewed (changes requested):
Some changes to be done, otherwise looks good to me.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 0f2ee10af012aa11b35eef65888d04d6ff9ea5ba).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87188872
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

xbauch commented:
The second commit rewrites the functionality following #918 style to simplify adding further checks.

NlightNFotis commented on src/goto-programs/goto_program.cpp:
```diff
@@ -668,6 +668,65 @@ bool goto_programt::instructiont::equals(const instructiont &other) const
+
+  switch(type)
+  {
+  case GOTO:
```
I think this whole construct can be simplified, unless there's a reason not to? just keep the `case ASSERT` and `case FUNCTION_CALL` that require special handling, and just remove the other ones and add one `default`?

xbauch replied:
There's method in this madness. When we add further checks we simply put them under the right case, without having to add more cases. (At least I think so: this part I've also copied from #918.)

NlightNFotis replied:
Another thing to keep in mind is the semantics of what you've written. The behaviour is the same for all of `case GOTO`, `case ASSUME` and `case ASSERT` for example, where for each one of them `guard.visit(symbol_finder);` will run because of fall through of the switch cases. Is that intended, or only `case ASSERT` should demonstrate that? If yes that's okay. If no, then what I suggested is a better/cleaner idea, and semantically correct while reducing the possibility of having cases being unintendedly massed together.

NlightNFotis reviewed (approved):
LGTM just one comment

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 30cc479a21001d3e837ed3e9f460b3bbaef5c4f5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87218803
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

sonodtt commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -516,6 +516,10 @@ int cbmc_parse_optionst::doit()
+  INVARIANT(
+    !goto_model.check_internal_invariants(*this),
+    "goto program internal invariants failed");
+
```
Not entirely sure what the best approach is, but this might be relevant going forward as we extend this.
I'm considering calling my goto_program checks from the try block of cbmc_parse_optionst::get_goto_program so that any exceptions of type 'incorrect_goto_program_exceptiont' are caught there.
If called directly from the doit() method exceptions are then caught in parse_options_baset::main()
which has only the base exception type or the invalid_user_input_exceptiont type. 
(I think)

xbauch replied:
Well in #3123 @danpoe makes it switchable to throw exception of check invariant. Also, the whole well-formedness is optional (user-wise). So, yes, I expect the point of entry to be modified at some point.

sonodtt replied:
Ok.

sonodtt reviewed (commented):
Comment was addressed.

danpoe commented on src/goto-programs/goto_program.cpp:
```diff
@@ -668,6 +668,65 @@ bool goto_programt::instructiont::equals(const instructiont &other) const
+  bool found_violation = false;
+  std::vector<std::string> id_collector;
+  auto symbol_finder = [&](const exprt &e) {
+    if(e.id() == ID_symbol)
```
I think we should also check the types with IDs `ID_symbol_type`, `ID_struct_tag`, `ID_c_enum_tag`, and `ID_union_tag` here (see `std_types.h`). Those types also hold an identifier which indexes into the symbol table.

xbauch replied:
Ok, I've added those checks and reworked the code to use `DATA_CHECK`.

danpoe commented on src/goto-programs/goto_program.cpp:
```diff
@@ -668,6 +668,65 @@ bool goto_programt::instructiont::equals(const instructiont &other) const
+    {
+      auto symbol_expr = to_symbol_expr(e);
+      if(!table.has_symbol(symbol_expr.get_identifier()))
+        id_collector.push_back(id2string(symbol_expr.get_identifier()));
```
We could probably also use the `DATA_CHECK()` macros here as in #3128. I was thinking if we want a third method where we keep going and report all errors we could add a third validation mode, say `validation_modet::KEEP_GOING` which does not throw an exception or yield a `DATA_INVARIANT` violation but instead prints the error messages to the `error()` stream. But that's future work I'd say.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: c6689853e807d73ae8254d184abc1062a8ddd722).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87512935

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 8cb487f048007a3f79d4524c4d23934b47de0daa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87618130

tautschnig commented on src/goto-programs/goto_function.h:
```diff
@@ -117,7 +117,59 @@ class goto_functiont
   void validate(const namespacet &ns, const validation_modet vm) const
   {
-    body.validate(ns, vm);
+    auto type_symbol_finder =
```
These seem very useful utilities - can we move them to/reconcile them with `util/find_symbols.{h,cpp}`?

xbauch replied:
Moved.

tautschnig commented on src/util/expr.cpp:
```diff
@@ -327,6 +327,24 @@ void exprt::visit(const_expr_visitort &visitor) const
   }
 }
 
+void exprt::visit(const std::function<void(const exprt &)> &f) const
```
Ok, but we should really have this in a separate commit, it would in fact warrant both documentation and a unit test.

xbauch replied:
Done.

tautschnig commented on unit/Makefile:
```diff
@@ -16,6 +16,7 @@ SRC += analyses/ai/ai.cpp \
        analyses/does_remove_const/is_type_at_least_as_const_as.cpp \
        compound_block_locations.cpp \
        goto-programs/goto_trace_output.cpp \
+       goto-programs/goto_program_table_consistency.cpp \
```
Please maintain lexicographic ordering.

tautschnig commented on unit/Makefile:
```diff
@@ -17,6 +17,7 @@ SRC += analyses/ai/ai.cpp \
        compound_block_locations.cpp \
        goto-programs/goto_trace_output.cpp \
        goto-programs/goto_program_table_consistency.cpp \
+       goto-programs/goto_model_function_type_consistency.cpp \
```
Lexicographic ordering, please.

xbauch replied:
Reordered.

xbauch commented:
I've added a few fixes to appease the CI, then squashed. The above comments should all be addressed.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0f33641ae68a304c0b80d5e67407cecbde51448f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91887744

xbauch commented:
@tautschnig I've rebased to include #3480 changes.

tautschnig commented:
@xbauch I'll take a look, but it seems CI is failing.

tautschnig commented on src/util/expr.cpp:
```diff
@@ -279,6 +279,20 @@ void exprt::visit(const_expr_visitort &visitor) const
+{
+  std::stack<const exprt *> stack;
+  stack.push(this);
+  while(!stack.empty())
```
Why is this not using the `depth_iteratort`? And I don't think it actually implements the visitor pattern in any way? Maybe we should at least not call this method `visit`?

xbauch replied:
I haven't noticed we have iterators (probably confused by the existing `visit` methods that do not use them). In which case I can as well call `std::for_each` directly.

xbauch replied:
And thus we don't need the unit test at all.

tautschnig replied:
Yay for less code!

tautschnig commented on src/util/expr.h:
```diff
@@ -305,6 +305,12 @@ class exprt:public irept
 
+  /// Recursively traverse the expression tree (via operands) and call \p f on
+  /// every subexpressions, including `this`. The input function has
+  /// side-effects and does not modify the expression.
```
"has _no_ side-effects" - or "may have side-effects"? At least I don't understand why we would have to prescribe that this function has side effects (though arguably it might not make sense to do this if it does not have any effect).

xbauch replied:
Changed to "may have side effects"

tautschnig commented on src/util/find_symbols.h:
```diff
@@ -14,10 +14,14 @@ Author: Daniel Kroening, kroening@kroening.com
 class symbol_exprt;
 class typet;
+class symbolt;
+class source_locationt;
```
Nit pick: lexicographic ordering.

tautschnig commented on src/goto-programs/goto_function.h:
```diff
@@ -118,6 +119,13 @@ class goto_functiont
   {
     body.validate(ns, vm);
+    source_locationt nil_location;
+    nil_location.make_nil();
```
Is this important?

xbauch commented:
@tautschnig Yes, my older tests were not smart enough for the new validation checks. Should be fine now.

tautschnig commented on src/goto-programs/goto_function.h:
```diff
@@ -118,6 +119,13 @@ class goto_functiont
+    forall_subtypes(it, type)
+    {
+      find_typetag_in_namespace(*it, nil_location, ns, vm);
+    }
```
Hmm, I don't get the impression that this is doing much of a "find." Couldn't you actually use one of the `find_*` functions to find the identifiers, and then check all of those that you have found?

xbauch replied:
@tautschnig Ok, but this last bit required a bit more code change so I pushed it as a new commit and will squash once approved.

tautschnig reviewed (commented):
Various smaller comments, but the part I'm most concerned about are the two new `find_*` functions that don't actually seem to be "finding" much, but just doing the checking. I think these are two distinct activities.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: bbeb69446c077e9655e3f9827b3d6de4b9afa890).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93516892
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

tautschnig reviewed (approved):
Looks good to me, especially the increased code reuse and less code being added.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 559e78fa644c00765b8bd02a4ed9790981a7e40f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93639702

Commit:
Symbol consistency between goto programs and symbol table

Iterate over all symbols in a goto program and for each one check that it is
present in the symbol table as well. Includes a unit test for the check.

Commit:
Function type consistency between goto programs and symbol table

A simple iteration over goto-functions to check that it's base type is the same
as the one stored in the symbol table. Includes a unit test.

Commit:
Program table symbol type consistency

Look up the symbol id in symbol table and call base_type_eq on every symbol
expression in guard and code whenever relevant. Includes a unit test.

Also fixes unit tests that these new checks brake.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 60a833e92779f45500abbd38234cd580b09eb901).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93651044
